core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -614,12 +614,24 @@ public boolean removeInferredStatement(Resource subj, IRI pred, Value obj, Resou
 
 	private boolean remove(Resource subj, IRI pred, Value obj, SailDataset dataset, SailSink sink, Resource... contexts)
 			throws SailException {
+
+//		(subj == null || pred == null || obj == null) &&
+		if (!hasConnectionListeners()
+				&& sink instanceof SailSinkVersion2 && getIsolationLevel().isCompatibleWith(IsolationLevels.NONE)) {
+			return ((SailSinkVersion2) sink).deprecateByQuery(subj, pred, obj, contexts);
+		}
```

Comment:
Going to need more performance testing at some point to see if deprecateByQuery is faster.